CLIENT_SPECIFICATION.md

```diff
@@ -0,0 +1,194 @@
+
+This document contains the official specification for tldr-pages clients. It is _not_ a specification of the format of the pages themselves - only a specification of how a user should be able to interface with an official client. 
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+
+## Changelog
+ - v1.0 (18th December 2018)
+	 - Initial draft
+ - v1.1 (18th December 2018)
+	 - Refined bullet points according to comments
+	 - Ensure the spec specifies the _interface_, not the workings
+ - v2.0 (12th January 2019)
+	 - Expanded notes into full sentences
+
+
+## Terminology
+In order to aid the understanding of this specification document, a number of terms will be defined in this section.
+
+### Page
+tldr-pages consists of multiple _pages_ - each of which describes a specific command.
+
+### Platform
+Pages are grouped by platform. A platform is an operating system. For example, `windows`, `linux`, `osx`. The special platform `common` contains pages for commands that work identically across all platforms.
+
+
+## Command-Line Interface
+This section describes the standardised command-line interface (CLI) for clients implementing one. Clients that do not provide a CLI can ignore this section.
+
+### Arguments
+A number of arguments MUST be supported if a CLI is implemented:
+
+Argument		        | Meaning
+------------------------|--------------------
+`--update`				| Updates the offline cache of pages.
+`--version`, `-v`		| Shows the current version of the client.
+`--list-all`, `-a`		| Lists all the pages in the current platform to the standard output. One page name per line. Additional decoration MAY be printed to the standard error.
+
+Clients MAY support additional custom arguments not documented here.
+
+Here are some examples invocations using the above flags:
+
+```bash
+tldr --update
+tldr --version
+tldr -a
+```
+
+### Page Names
+The first argument that does not start with a dash (`-`) MUST be considered the page name.
+
+In addition, page names MAY contain spaces (e.g. `git status`) - such page names MUST be transparently concatenated with dashes (`-`). For example, the page name:
+
+```
+git checkout
+```
+
+...becomes this:
+
+```
+git-checkout
+```
+
+Here are some example invocations:
+
+```bash
+tldr 7za
+tldr eyeD3
+tldr git checkout
+# In the below, "--foo" is a custom argument that takes a parameter.
+tldr --foo bar bash
+```
+
+#### Specifying the Platform
+As pages are grouped by platform, a user may want to access a platform-specific version of a page. This MUST be supported by prefixing the page name as follows:
+
+```
+windows/type
+linux/sh
+osx/brew
+```
+
+Example invocations are as follows:
+
+```bash
+tldr windows/type
+# --foo is a custom argument, as described above
+tldr --foo bar common/git merge
```

Comment:
I have never seen a commandline application that uses `/` as anything but a path separator. I teach an introduction to the commandline class and it will be very confusing for students to see this formatting. It is also telling that this kind of formatting is rarely supported well by cli argument parsing libraries.

I am not sure if this is the best place to argue for a commandline option to disambiguate via `--os`. I feel like it will be similar to `--language`.

Reply:
(and I've seen clients that allow you to just give a path to a markdown file to render it)

Reply:
It could be a good idea to allow something like `tldr platform command` like `man` does with `man section command` :)

Reply:
so the only tricky thing there is the spec mentions `tldr platform command` should be disambiguated as `tldr platform-command`. That said, matching `linux|macos|winndows` wouldnt be so difficult.

Reply:
And then you have following issue. What's expected result of this hypothetical case: `tldr osx cleaner`. Is it `cleaner` app tldr page on `osx` platform or simply an app named `osx cleaner` with `osx-cleaner` tldr page? 🤔

Reply:
Your proposal of `tldr PLATFORM PAGE NAME` syntax could work, because so far there are no pages that start with any of the platforms that are supported. It would be even easier to type, and take care of @jedahan's objections. I initially chose `/` because in the current archive implementation, it does boil down to being a path separator (and it is shorter to type than some --prefix and easier to remember). In the current tldr-bash-client I could easily start supporting both forms. (This won't really help the client specification though..!)

Reply:
(By the way, the syntax to specify the path to a markdown file for rendering does require a commandline flag, and doesn't clash with the use of `/` in a normal page lookup.)

Reply:
I see. The pages do actually have 1 folder per platform though. Also, I see the `/` as a 'category' separator, sort-of meaning "inside".

So my typing `tldr linux/tar`, I'm requesting the `tar` page in the `linux` platform.

Reply:
What is the issue with having a flag for platform, like `--platform=osx` ? Seems unambiguous to me.

Reply:
The thing is that `tldr platform/page` requires much less typing that `--platform=os_name`

Reply:
By going with forward slashes it becomes path-like so it is hard to guess what's really happening here (for newcomers) as @jedahan already pointed out. Also it will look superweird on windows because of forward slashes. If you really want to avoid flags that much use something less ambigous/extravagant like `tldr platform:page` :bulb:

Reply:
Hmm .. IMO I think the amount of time saved by not typing a few characters is not worth the confusion caused due to it.

Reply:
`--platform, -p` or `--os, -o` if its kinda short.
I set this anyway because its usually what I want: `h() { tldr -q $* || tldr -q -o linux $* || tldr -q -o windows $* }`

Reply:
In that case, we need to decide between `--platform` or `--os`. I do like `platform:page`  though.

Reply:
Lets make sure that we all are on same page. @agnivade pointed out that saving few keystrokes wont save a day comparing to possible confusion it could create and I fully agree with that one. In the same time @sbrl pointed out that `platform/page` is less tedious to write than `--platform=<os_name>` to which I can totally relate. @jedahan introduced short options into the story as some kind of compromise?

Folks, few things to have in mind:
1. We created discussion around long options while in reality it goes other way around.
Original UNIX had only short options and those long ones came with GNU tools as a way of expanding possible options pool. Rule of thumb is that every option should have short flag while optionally long one can be added for user ergonomics. That's not some dogmatic rule that's [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html).
OFC that experienced users (familiar with particular tool) would favor `-p <platform>` over `--platform=<platform>` while latter provides more information to newcomers. That's not my prediction, that's [GNU's message to developers](https://www.gnu.org/prep/standards/html_node/Command_002dLine-Interfaces.html).
At the end of the day who doesn't want to be both POSIX and GNU compliant and that is why most of well known tools these days support both short & long options.

2. @agnivade in his replies always used `--platform=<platform>` form.
Personally, that's worst (or at least most tedious) form of GNU style one could use; let me explain way. Aside from `--` and _long option_ itself you also use `=` which is located on _edge of typing space_ on standard keyboard.
I hope this isn't news for anyone but just to make sure: [relevant section of TAO of Unix Programming](http://www.catb.org/esr/writings/taoup/html/ch10s05.html) quoted here:

    > An option argument (if any) can be separated by either whitespace or a single “=” (equal sign) character.

    Which means you could use `--platform <platform>`.

    As meme says __I don't usually do GNU style options but when I do I use whitespace instead.__ You can probably guess my rationale, yeah it is easier to hit on keyboard. That's just my personal preference, you may or may not find it true in your case; I just wanted to point that out.

    **GNU gives you choice**:
    - `--platform=<platform>` :white_check_mark:
    - `--platform <platform>` :white_check_mark:

    Pick your poison :wink:

3. Regarding `--os` vs `--platform` and `-o` vs `-p`.

    First pair is debating between concepts. Is it really an OS or a platform? I would say neither but in my previous reply https://github.com/tldr-pages/tldr/pull/2706#discussion_r I explained why my vote goes to `--platform`.

    Second pair is debating between expectations: [standard options](http://www.catb.org/esr/writings/taoup/html/ch10s05.html). My rationale here is that `-p` as _port_ makes sense only in context of networking tools while `-o` as _output_ applies to more wider variety of usecases. So my vote here goes to `-p`.

    `--platform` + `-p` = `-p, --platform`

## Proposal

Clients **MUST** implement `-p, --platform`. That covers both, ergonomic needs of experienced users and learning needs of newcomers and it is POSIX and GNU compliant. Clients **MAY** additionally support `platform:page` shorthand in lieu of regular options to provide both experienced and new users with expressive and intuitive query syntax.

Thoughts? :speaking_head:

Reply:
To further support my case lets examine what @mebeim brought to discussion: https://github.com/tldr-pages/tldr/pull/2706#discussion_r247359500:
>It could be a good idea to allow something like `tldr platform command` like man does with `man section command` :) Sounds pretty straightforward to me.

According to [`man man`](https://www.freebsd.org/cgi/man.cgi?query=man&apropos=0&sektion=0&manpath=FreeBSD+12.0-RELEASE+and+Ports&arch=default&format=html) itself:
```
SYNOPSIS
     man [-adho] [-t | -w] [-M manpath] [-P pager] [-S mansect]
     [-m arch[:machine]] [-p [eprtv]] [mansect] page ...
     man -f keyword ...
     man -k keyword ...
```

Optional `[mansect]` argument is just convenience shorthand for `[-S mansect]`. Take for example:
```
$ man 1 printf
$ man 3 printf
```
those are just shortcuts for:
```
$ man printf -S1
$ man printf -S3
```
_(FWIW `-S` is actually a colon delimited range so `-S1` and `-S1:2` would produce same result.)_

or if you hate hitting that <kbd>Shift</kbd> key you can leverage System V compatibility through:
```
$ man printf -s1
$ man printf -s3
```
Strangely enough this is not even listed as possible but simply implemented behind curtains.

Man sections are not really analogous to platforms but there is another interesting option here:
```
[-m arch[:machine]
```
It enables overriding `MACHINE` and `MACHINE_ARCH` environment variables _(e.g. `MACHINE=macppc MACHINE_ARCH=powerpc` for those old PowerPC powered Macs)_ in order to load **hardware platform** specific manpages. This was probably less exotic back at the time while currently it does not serve much of a purpose.
Take a closer look on syntax tho: `arch:machine` _(e.g. macppc:powerpc for my previous example)_ I knew that someone implanted those archaic ideas in my hand :angry:

>**This gives us a history lesson about following standards and extending them with and unique expressive bits and pieces.**

Reply:
In the bash-tldr client using the `tldr platform/page` syntax is standard. I am happy to support `tldr platform:page` as well, and would be willing to also implement `tldr platform page`. I might go all out and support whatever commandline options are going to be mandated -- as long as they don't conflict of course.

Reply:
Thanks for the excellent analysis, @vladimyr! Here are my thoughts:

 - I think `tldr platform page name` is slightly less clear on which bit is the platform and which the page.
 - I agree on the subject of os vs platform. We might support something like `android` as a platform in the future for example - which is technically linux as an OS, but a completely different platform with a completely different set of commands.
 - I like the `-p page`, `--platform page` proposal a lot.
 - I like `platform:page` a lot too.

I agree with @vladimyr's proposal for MUST `-p`, `--platform` and MAY `platform:page`.